File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/mad/Java/Go/s514557145_comments_removed.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"math/bits"
	"os"
)

const (
	MOD1 = 1000000007
	MOD2 = 998244353
)

var (
	S_E = S{0, (1<<31 - 1)/2, -(1 << 31)/2}
	F_Id = F{0, 0}
)

type S struct {
	Sum int64
	L   int
	R   int
}

type F struct {
	D int64
	N int
}

func modInv(x int64) int64 {
	return modPow(x, MOD2-2)
}

func modPow(x, y int64) int64 {
	z := int64(1)
	for y > 0 {
		if y%2 == 0 {
			x = (x * x) % MOD2
			y /= 2
		} else {
			z = (z * x) % MOD2
			y--
		}
	}
	return z
}

var inv9 = modInv(9)

func opS(a, b S) S {
	sum := a.Sum + b.Sum
	if sum >= MOD2 {
		sum -= MOD2
	}
	newL := a.L
	if b.L < newL {
		newL = b.L
	}
	newR := a.R
	if b.R > newR {
		newR = b.R
	}
	return S{sum, newL, newR}
}

func mapS(f F, s S) S {
	n := f.N
	tenmodL := tenmod[n-s.L]
	var tenmodR int64
	if s.R == n-1 {
		tenmodR = 1 // since tenmod[0] is ScrollBar
	} else {
		tenmodR = tenmod[n-s.R-1]
	}
	c := (tenmodL - tenmodR) % MOD2
	if c < 0 {
		c += MOD2
	}
	sum := (f.D * c) % MOD2
	sum = (sum * inv9) % MOD2
	return S{sum: sum, L: s.L, R: s.R}
}

func compositeF(f, g F) F {
	return F{D: f.D, N: f.N}
}

type LazySegTree struct {
	Dat     []S
	Laz     []F
	N       int
	Log     int
	Op      func(a, b S) S
	E       S
	Mapping func(f F, s S) S
	Composition func(f, g F) F
	Id      F
	Max     int
}

func NewLazySegTree(data []S, op func(a, b S) S, e S, mapping func(f F, s S) S, composition func(f, g F) F, id F) *LazySegTree {
	n := len(data)
	size := 1
	for size < n {
		size <<= 1
	}
	log := bits.TrailingZeros(uint(size))
	dat := make([]S, 2*size)
	for i := range dat {
		dat[i] = e
	}
	copy(dat[size:], data)
	for i := size - 1; i > 0; i-- {
		dat[i] = op(dat[2*i], dat[2*i+1])
	}
	laz := make([]F, size)
	for i := range laz {
		laz[i] = id
	}
	return &LazySegTree{
		Dat: dat,
		Laz: laz,
		N: size,
		Log: log,
		Op: op,
		E: e,
		Mapping: mapping,
		Composition: composition,
		Id: id,
		Max: n,
	}
}

func (t *LazySegTree) push(k int) {
	if t.Laz[k] == t.Id {
		return
	}
	l := 2 * k
	r := 2*k + 1
	t.Dat[l] = t.Mapping(t.Laz[k], t.Dat[l])
	t.Dat[r] = t.Mapping(t.Laz[k], t.Dat[r])
	if l < t.N {
		t.Laz[l] = t.Composition(t.Laz[k], t.Laz[l])
	}
	if r < t.N {
		t.Laz[r] = t.Composition(t.Laz[k], t.Laz[r])
	}
	t.Laz[k] = t.Id
}

func (t *LazySegTree) pushTo(k int) {
	for i := t.Log; i > 0; i-- {
		t.push(k >> i)
	}
}

func (t *LazySegTree) pushToRange(l, r int) {
	for i := t.Log; i > 0; i-- {
		if ((l >> i) << i) != l {
			t.push(l >> i)
		}
		if ((r >> i) << i) != r {
			t.push(r >> i)
		}
	}
}

func (t *LazySegTree) updateFrom(l, r int) {
	for i := 1; i <= t.Log; i++ {
		if ((l >> i) << i) != l {
			lki := l >> i
			left := lki * 2
			right := left + 1
			t.Dat[lki] = t.Op(t.Dat[left], t.Dat[right])
		}
		if ((r >> i) << i) != r {
			rki := (r - 1) >> i
			left := rki * 2
			right := left + 1
			t.Dat[rki] = t.Op(t.Dat[left], t.Dat[right])
		}
	}
}

func (t *LazySegTree) apply(l, r int, f F) {
	if l >= r {
		return
	}
	l += t.N
	r += t.N
	t.pushToRange(l, r)
	for l < r {
		if l&1 == 1 {
			t.Dat[l] = t.Mapping(f, t.Dat[l])
			if l < t.N {
				t.Laz[l] = t.Composition(f, t.Laz[l])
			}
			l++
		}
		if r&1 == 1 {
			r--
			t.Dat[r] = t.Mapping(f, t.Dat[r])
			if r < t.N {
				t.Laz[r] = t.Composition(f, t.Laz[r])
			}
		}
		l >>= 1
		r >>= 1
	}
	t.updateFrom(l, r)
}

func (t *LazySegTree) allProd() S {
	return t.Dat[1]
}

type Reader struct {
	r *bufio.Reader
}

func NewReader(r io.Reader) *Reader {
	return &Reader{r: bufio.NewReader(r)}
}

func (rd *Reader) readByte() (byte, error) {
	return rd.r.ReadByte()
}

func (rd *Reader) skipSpace() {
	for {
		b, _ := rd.readByte()
		if b == ' ' || b == '	' || b == '
' || b == '' {
			continue
		} else {
			rad.r.UnreadByte()
			break
		}
	}
}

func (rd *Reader) readInt() int {
	rad.skipSpace()
	sign := 1
	if b, _ := rd.readByte(); b == '-' {
		sign = -1
	} else {
	rad.r.UnreadByte()
	}
	num := 0
	for {
		b, err := rd.readByte()
		if err != nil || b < '0' || b > '9' {
			break
		}
		num = num*10 + int(b - '0')
	}
	return sign * num
}

func (rd *Reader) readLong() int64 {
	rad.skipSpace()
	sign := int64(1)
	if b, _ := rd.readByte(); b == '-' {
		sign = -1
	} else {
	rad.r.UnreadByte()
	}
	num := int64(0)
	for {
		b, err := rd.readByte()
		if err != nil || b < '0' || b > '9' {
			break
		}
		num = num*10 + int64(b - '0')
	}
	return sign * num
}

var tenmod []int64

func main() {
	N := NewReader(os.Stdin).readInt()
	Q := NewReader(os.Stdin).readInt()
	tenmod = make([]int64, N+1)
	tenmod[0] = 1
	for i := 1; i <= N; i++ {
		tenmod[i] = (tenmod[i-1] * 10) % MOD2
	}
	data := make([]S, N)
	for i := 0; i < N; i++ {
		data[i] = S{
			Sum: tenmod[N-i-1],
			L:   i,
			R:   i,
		}
	}
	st := NewLazySegTree(data, opS, S_E, mapS, compositeF, F_Id)
	for i := 0; i < Q; i++ {
		rad := NewReader(os.Stdin)
		l := rad.readInt() - 1
		r := rad.readInt() - 1
		D := rad.readLong()
		st.apply(l, r+1, F{D: D, N: N})
		fmt.Println(st.allProd().Sum)
	}
} // End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/mad/Java/Python/s514557145_comments_removed.py
class S:
    __slots__ = ['sum', 'l', 'r']
    def __init__(self, sum_val, l, r):
        self.sum = sum_val % MOD2
        self.l = l
        self.r = r

    @classmethod
    def combine(cls, a, b, mod):
        new_sum = (a.sum + b.sum) % mod
        new_l = min(a.l, b.l)
        new_r = max(a.r, b.r)
        return S(new_sum, new_l, new_r)

    @classmethod
    def map_f(cls, s, f, tenmod, N, inv9, mod):
        c = (tenmod[N - s.l] - tenmod[N - s.r - 1]) % mod
        new_sum = (f.D * c % mod) * inv9 % mod
        return S(new_sum, s.l, s.r)

class F:
    __slots__ = ['D', 'N']
    def __init__(self, D, N):
        self.D = D
        self.N = N

class LazySegTree:
    def __init__(self, data, tenmod, N, inv9, MOD):
        self.MOD = MOD
        self.n = len(data)
        self.size = 1
        while self.size < self.n:
            self.size <<= 1
        self.log = self.size.bit_length() - 1
        self.dat = [S(0, 0, 0) for _ in range(2 * self.size)]
        self.laz = [None for _ in range(self.size)]
        for i in range(self.n):
            self.dat[self.size + i] = data[i]
        for i in range(self.size - 1, 0, -1):
            self.dat[i] = S.combine(self.dat[i << 1], self.dat[i << 1 | 1], self.MOD)
        self.tenmod = tenmod
        self.N = N
        self.inv9 = inv9

    def _apply(self, node, f):
        s = self.dat[node]
        new_s = S.map_f(s, f, self.tenmod, self.N, self.inv9, self.MOD)
        self.dat[node] = new_s
        if node < self.size:
            self.laz[node] = f

    def _push(self, node):
        if self.laz[node]:
            f = self.laz[node]
            self._apply(node << 1, f)
            self._apply(node << 1 | 1, f)
            self.laz[node] = None

    def _push_all_ancestors(self, node):
        current = node
        while current > 0:
            self._push(current)
            current >>= 1

    def apply_range(self, l, r, f):
        l += self.size
        r += self.size
        self._push_all_ancestors(l)
        self._push_all_ancestors(r - 1)
        l0, r0 = l, r
        while l0 < r0:
            if l0 & 1:
                self._apply(l0, f)
                l0 += 迷 1
            if r0 & 1:
                r0 -= 1
                self._apply(r0, f)
            l0 >>= 1
            r0 >>= 1

    def get_total_sum(self):
        return self.dat[1].sum % self.MOD

def main():
    import sys
    MOD2 = 998244353
    inv9 = pow(9, MOD2-2, MOD2)
    input = sys.stdin.read().split()
    ptr = 0
    N = int(input[ptr]); ptr +=1
    Q = int(input[ptr]); ptr +=1
    tenmod = [1]*(N+1)
    for j in range(1, N+1):
        tenmod[j] = (tenmod[j-1] * 10) % MOD2
    data = []
    for i in range(N):
        coeff = tenmod[N - i -1]
        data.append(S(coeff, i, i))
    seg_tree = LazySegTree(data, tenmod, N, inv9, MOD2)
    for _ in range(Q):
        l = int(input[ptr])-1; ptr +=1
        r = int(input[ptr])-1; ptr +=1
        D = int(input[ptr]); ptr +=1
        f = F(D, N)
        seg_tree.apply_range(l, r+1, f)
        print(seg_tree.get_total_sum())

if __name__ == "__main__":
    main()  # End of Code
